src/bt-hh-io.c

```diff
@@ -0,0 +1,714 @@
+  if ((append_bt_bdaddr_t(&wbuf->buf.pdu, bd_addr) < 0) ||
+      (append_to_pdu(&wbuf->buf.pdu, "C", (uint8_t)state) < 0))
+    goto cleanup;
+
+  if (run_task(send_ntf_pdu, wbuf) < 0)
+    goto cleanup;
+
+  return;
+cleanup:
+  cleanup_pdu_wbuf(wbuf);
+}
+
+static void
+hid_info_cb(bt_bdaddr_t* bd_addr, bthh_hid_info_t hid_info)
+{
+  struct pdu_wbuf* wbuf;
+
+  wbuf = create_pdu_wbuf(6 + /* address */
+                         27 + hid_info.dl_len, /* hid info */
+                         0, NULL);
+
+  if (!wbuf)
+    return;
+
+  init_pdu(&wbuf->buf.pdu, SERVICE_BT_HH, OPCODE_HID_INFO_NTF);
+  if ((append_bt_bdaddr_t(&wbuf->buf.pdu, bd_addr) < 0) ||
+      (append_bthh_hid_info_t(&wbuf->buf.pdu, &hid_info) < 0))
+    goto cleanup;
+
+  if (run_task(send_ntf_pdu, wbuf) < 0)
+    goto cleanup;
+
+  return;
+cleanup:
+  cleanup_pdu_wbuf(wbuf);
+}
+
+static void
+protocol_mode_cb(bt_bdaddr_t* bd_addr, bthh_status_t hh_status,
+  bthh_protocol_mode_t mode)
+{
+  struct pdu_wbuf* wbuf;
+
+  wbuf = create_pdu_wbuf(6 + /* address */
+                         1 + /* HH status */
+                         1, /* protocol mode */
+                         0, NULL);
+
+  if (!wbuf)
+    return;
+
+  init_pdu(&wbuf->buf.pdu, SERVICE_BT_HH, OPCODE_PROTOCOL_MODE_NTF);
+  if ((append_bt_bdaddr_t(&wbuf->buf.pdu, bd_addr) < 0) ||
+      (append_to_pdu(&wbuf->buf.pdu, "CC", (uint8_t)hh_status,
+                                           (uint8_t)mode) < 0))
+    goto cleanup;
+
+  if (run_task(send_ntf_pdu, wbuf) < 0)
+    goto cleanup;
+
+  return;
+cleanup:
+  cleanup_pdu_wbuf(wbuf);
+}
+
+static void
+idle_time_cb(bt_bdaddr_t* bd_addr, bthh_status_t hh_status, int idle_rate)
+{
+  struct pdu_wbuf* wbuf;
+
+  wbuf = create_pdu_wbuf(6 + /* address */
+                         1 + /* HH status */
+                         4, /* idle rate */
+                         0, NULL);
+
+  if (!wbuf)
+    return;
+
+  init_pdu(&wbuf->buf.pdu, SERVICE_BT_HH, OPCODE_IDLE_TIME_NTF);
+  if  ((append_bt_bdaddr_t(&wbuf->buf.pdu, bd_addr) < 0) ||
+       (append_to_pdu(&wbuf->buf.pdu, "CS", (uint8_t)hh_status,
+                                            (uint16_t)idle_rate) < 0))
+    goto cleanup;
+
+  if (run_task(send_ntf_pdu, wbuf) < 0)
+    goto cleanup;
+
+  return;
+cleanup:
+  cleanup_pdu_wbuf(wbuf);
+}
+
+static void
+get_report_cb(bt_bdaddr_t* bd_addr, bthh_status_t hh_status, uint8_t* rpt_data,
+  int rpt_size)
+{
+  struct pdu_wbuf* wbuf;
+
+  wbuf = create_pdu_wbuf(6 + /* address */
+                         1 + /* HH status */
+                         4 + /* report size */
+                         rpt_size, /* report data */
+                         0, NULL);
+
+  if (!wbuf)
+    return;
+
+  init_pdu(&wbuf->buf.pdu, SERVICE_BT_HH, OPCODE_GET_REPORT_NTF);
+  if ((append_bt_bdaddr_t(&wbuf->buf.pdu, bd_addr) < 0) ||
+      (append_to_pdu(&wbuf->buf.pdu, "CSm", (uint8_t)hh_status,
+        (uint16_t)rpt_size, rpt_data, rpt_size) < 0))
```

Comment:
rpt_size is correct here.

Please align values below each other. rpt_size must be cast to size_t to serve as array length for 'm'